packages/cli/src/processed-data-managers/processed-data-helper.ts

```diff
@@ -0,0 +1,322 @@
+			if (options.maxEntries) {
+				hashedItems.splice(0, hashedItems.length - options.maxEntries);
+			}
+			await Container.get(ProcessedDataRepository).insert({
+				workflowId: contextData.workflow.id.toString(),
+				context: dbContext,
+				value: {
+					mode: options.mode,
+					data: hashedItems,
+				},
+			});
+
+			return {
+				new: items,
+				processed: [],
+			};
+		}
+
+		const returnData: ICheckProcessedOutput = {
+			new: [],
+			processed: [],
+		};
+
+		const processedDataValue = processedData.value as IProcessedDataEntries;
+
+		hashedItems.forEach((item, index) => {
+			if (processedDataValue.data.find((entry) => entry === item)) {
+				returnData.processed.push(items[index]);
+			} else {
+				returnData.new.push(items[index]);
+				processedDataValue.data.push(item);
+			}
+		});
+
+		if (options.maxEntries) {
+			processedDataValue.data.splice(0, processedDataValue.data.length - options.maxEntries);
+		}
+
+		await Container.get(ProcessedDataRepository).save(processedData);
+
+		return returnData;
+	}
+
+	async removeProcessed(
+		items: ProcessedDataItemTypes[],
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+		options: ICheckProcessedOptions,
+	): Promise<void> {
+		if (['latestIncrementalKey', 'latestDate'].includes(options.mode)) {
+			throw new ApplicationError('Removing processed data is not possible in mode "latest"');
+		}
+
+		const processedData = await Container.get(ProcessedDataRepository).findOne({
+			where: {
+				workflowId: contextData.workflow.id as string,
+				context: ProcessedDataHelper.createContext(context, contextData),
+			},
+		});
+
+		if (!processedData) {
+			return;
+		}
+
+		const hashedItems = items.map((item) => ProcessedDataHelper.createValueHash(item));
+
+		const processedDataValue = processedData.value as IProcessedDataEntries;
+
+		hashedItems.forEach((item) => {
+			const index = processedDataValue.data.findIndex((value) => value === item);
+			if (index !== -1) {
+				processedDataValue.data.splice(index, 1);
+			}
+		});
+
+		await Container.get(ProcessedDataRepository).save(processedData);
+	}
+
+	async clearAllProcessedItems(
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+		options: ICheckProcessedOptions,
+	): Promise<void> {
+		console.log(options);
+		await Container.get(ProcessedDataRepository).delete({
+			workflowId: contextData.workflow.id as string,
+			context: ProcessedDataHelper.createContext(context, contextData),
+		});
+	}
+
+	async getProcessedDataCount(
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+		options: ICheckProcessedOptions,
+	): Promise<number> {
+		const processedDataRepository = Container.get(ProcessedDataRepository);
+
+		const processedData = await processedDataRepository.findOne({
+			where: {
+				workflowId: contextData.workflow.id as string,
+				context: ProcessedDataHelper.createContext(context, contextData),
```

Comment:
This is a bit confusing as we are creating a context but we are passing a context to it as a parameter. If `ProcessedDataContext` is `'node' | 'workflow'`, could the "context" that we are creating here be called for example `contextKey` or something like that?

Reply:
totally I can see how confusing this is, I renamed it to scope